pkg/apiserver/rest/dualwriter_syncer.go

```diff
@@ -0,0 +1,285 @@
+
+			item, ok := itemsByName[name]
+			if !ok {
+				item = syncItem{}
+			}
+			item.name = name
+			item.objLegacy = obj
+			itemsByName[name] = item
+		}
+
+		for _, obj := range storageList {
+			accessor, err := utils.MetaAccessor(obj)
+			if err != nil {
+				log.Error(err, "error retrieving accessor data for object from storage")
+				continue
+			}
+			name := accessor.GetName()
+
+			item, ok := itemsByName[name]
+			if !ok {
+				item = syncItem{}
+			}
+			item.name = name
+			item.objStorage = obj
+			itemsByName[name] = item
+		}
+		log.Info("got list of items to be synced", "items", len(itemsByName))
+
+		for name, item := range itemsByName {
+			// upsert if:
+			// - existing in both legacy and storage, but objects are different, or
+			// - if it's missing from storage
+			if item.objLegacy != nil &&
+				((item.objStorage != nil && !Compare(item.objLegacy, item.objStorage)) || (item.objStorage == nil)) {
+				outOfSync++
+
+				accessor, err := utils.MetaAccessor(item.objLegacy)
+				if err != nil {
+					log.Error(err, "error retrieving accessor data for object from storage")
+					continue
+				}
+
+				if item.objStorage != nil {
+					accessorStorage, err := utils.MetaAccessor(item.objStorage)
+					if err != nil {
+						log.Error(err, "error retrieving accessor data for object from storage")
+						continue
+					}
+					accessor.SetResourceVersion(accessorStorage.GetResourceVersion())
+					accessor.SetUID(accessorStorage.GetUID())
+
+					log.Info("updating item on unified storage", "name", name)
+				} else {
+					accessor.SetResourceVersion("")
+					accessor.SetUID("")
+
+					log.Info("inserting item on unified storage", "name", name)
+				}
+
+				objInfo := rest.DefaultUpdatedObjectInfo(item.objLegacy, []rest.TransformFunc{}...)
+				res, _, err := storage.Update(ctx,
+					name,
+					objInfo,
+					func(ctx context.Context, obj runtime.Object) error { return nil },
+					func(ctx context.Context, obj, old runtime.Object) error { return nil },
+					true, // force creation
+					&metav1.UpdateOptions{},
+				)
+				if err != nil {
+					log.WithValues("object", res).Error(err, "could not update in storage")
+					syncErr++
+				} else {
+					syncSuccess++
+				}
+			}
+
+			// delete if object does not exists on legacy but exists on storage
+			if item.objLegacy == nil && item.objStorage != nil {
+				outOfSync++
+
+				ctx = request.WithRequestInfo(ctx, &request.RequestInfo{
+					APIGroup:  requestInfo.APIGroup,
+					Resource:  requestInfo.Resource,
+					Name:      name,
+					Namespace: requestInfo.Namespace,
+				})
+
+				log.Info("deleting item from unified storage", "name", name)
+
+				deletedS, _, err := storage.Delete(ctx, name, func(ctx context.Context, obj runtime.Object) error { return nil }, &metav1.DeleteOptions{})
+				if err != nil {
+					if !apierrors.IsNotFound(err) {
+						log.WithValues("objectList", deletedS).Error(err, "could not delete from storage")
+					}
+					syncErr++
```

Comment:
[doubt] Do we want to consider the not found error as a sync error here? This would be the case of things changing while we sync, right? If that's the case, IMO we could consider this case a success since the element is now in sync (even if it wasn't us who deleted it), and thus report it as a sync success.

Reply:
you are right!

I changed the code to reflect that